support/src/main/java/org/jenkinsci/plugins/workflow/support/steps/StageStepExecution.java

```diff
@@ -222,18 +222,26 @@ private static void cleanUp(Job<?,?> job, String jobName) {
     }
 
     private static void println(StepContext context, String message) {
+        if (!context.isReady()) {
+            LOGGER.log(Level.FINE, "cannot print message ‘{0}’ to dead {1}", new Object[] {message, context});
+            return;
+        }
         try {
             context.get(TaskListener.class).getLogger().println(message);
         } catch (Exception x) {
-            LOGGER.log(WARNING, null, x);
+            LOGGER.log(WARNING, "failed to print message to dead " + context, x);
         }
     }
 
     // TODO record the stage it got to and display that
     private static void cancel(StepContext context, StepContext newer) throws IOException, InterruptedException {
-        println(context, "Canceled since " + newer.get(Run.class).getDisplayName() + " got here");
-        println(newer, "Canceling older " + context.get(Run.class).getDisplayName());
-        context.onFailure(new FlowInterruptedException(Result.NOT_BUILT, new CanceledCause(newer.get(Run.class))));
+        if (context.isReady() && newer.isReady()) {
+            println(context, "Canceled since " + newer.get(Run.class).getDisplayName() + " got here");
+            println(newer, "Canceling older " + context.get(Run.class).getDisplayName());
+            context.onFailure(new FlowInterruptedException(Result.NOT_BUILT, new CanceledCause(newer.get(Run.class))));
+        } else {
+            LOGGER.log(WARNING, "cannot cancel dead {0} or {1}", new Object[] {context, newer});
```

Comment:
So in this case, the older build will no be cancelled and the new one will proceed without any confirmation of the old one being killed. I think this could lead to weird situations if the old build comes back to live sometime... but unfortunately I don't see any _secure_ fix for this corner case.